### ISSUE TITLE
Failed to unserialize data

### ISSUE THREAD

**Issue by basz** [php-ds/extension#77]
Hi,

I'm using php-ds quite a bit. In our event sourced system de/serialization is done on snapshots of aggregates. After installing the php-ds extension I suddenly experience deserialisation errors.

Not sure where to begin, but it is probably related to Set or Map

https://gist.github.com/basz/c780c13b5a1b9c9e40a06d4b9d3f037b

```

$ APP_ENV=development php ./bin/test.php 
Failed to unserialize data
#0 /home-projects/api-plhw-development/deploy/releases/20170210133916UTC/bin/test.php(41): unserialize('O:47:"HF\\Model\\...')
#1 {main}
```

I'm using : 
php-ds is already the newest version (1.1.7-1+deb.sury.org~xenial+2).
php 7.1.2

**Comment by basz** [comment]
here is a failing test case

```
<?php

declare(strict_types=1);

class Identifier implements \Ds\Hashable
{
    private $identifier;

    public static function with(string $identifier): Identifier
    {
        $me             = new static();
        $me->identifier = $identifier;

        return $me;
    }

    public function hash()
    {
        return sha1($this->identifier);
    }

    /**
     * @param Identifier $obj
     * @return bool
     */
    public function equals($obj): bool
    {
        return get_class($this) == get_class($obj) &&
            $this->identifier === $obj->identifier;
    }
}

class SimpleClass
{
    private $map;

    public function __construct()
    {
        $this->map = new \Ds\Map();

        $this->map->put(Identifier::with('abc'), 'value');
    }
}

$simpleClass = new SimpleClass();

var_dump($simpleClass);

$serialized = serialize($simpleClass);

print $serialized;
try {
    $simpleClassDeserialized = unserialize($serialized);
} catch (\Throwable $e) {
    print $e->getMessage() . PHP_EOL;
    print $e->getTraceAsString() . PHP_EOL;
}

var_dump($simpleClassDeserialized);

```

```
object(SimpleClass)#3 (1) {
  ["map":"SimpleClass":private]=>
  object(Ds\Map)#2 (1) {
    [0]=>
    object(Ds\Pair)#5 (2) {
      ["key"]=>
      object(Identifier)#4 (1) {
        ["identifier":"Identifier":private]=>
        string(3) "abc"
      }
      ["value"]=>
      string(5) "value"
    }
  }
}
O:11:"SimpleClass":1:{s:16:"SimpleClassmap";C:6:"Ds\Map":74:{O:10:"Identifier":1:{s:22:"Identifieridentifier";s:3:"abc";}s:5:"value";}}Failed to unserialize data
#0 /home-projects/api-plhw-development/deploy/releases/20170210133916UTC/bin/test.php(81): unserialize('O:11:"SimpleCla...')
#1 {main}
NULL
```

**Comment by nikic** [comment]
Culprit: https://github.com/php-ds/extension/blob/88b292fba1826a58cd5b40f1789348416a15242b/src/ds/ds_htable.c#L1227

**Comment by rtheunissen** [comment]
Borrowed that from SPL structure code. Should I just drop that check @nikic? Will take a closer look this week.

**Comment by basz** [comment]
so relieved you all seem to know what's going on :-p. Thanks in advance!

**Comment by nikic** [comment]
@rtheunissen You should be checking against the provided `length` parameter. I'm pretty sure `unserialize` doesn't get a terminated string, so the null check will only work if a map is serialized directly (as opposed to being part of a structure).

**Comment by rtheunissen** [comment]
@nikic could you please check https://github.com/php-ds/extension/pull/80?

**Comment by rtheunissen** [comment]
Fixed and released as 1.1.8

**Comment by basz** [comment]
Thanks!
